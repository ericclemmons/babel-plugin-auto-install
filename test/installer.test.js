var expect = require("expect");
var fs = require("fs");
var path = require("path");
var spawn = require("cross-spawn");

var installer = require("../src/installer");

describe("installer", function() {
  describe(".defaultOptions", function() {
    it("should default dev to false", function() {
      expect(installer.defaultOptions.dev).toEqual(false);
    });

    it("should default peerDependencies to true", function() {
      expect(installer.defaultOptions.peerDependencies).toEqual(true);
    });
  })

  describe(".check", function() {
    context("given nothing", function() {
      it("should return undefined", function() {
        expect(installer.check()).toBe(undefined);
      });
    });

    context("given a local module", function() {
      it("should return undefined", function() {
        expect(installer.check("./foo")).toBe(undefined);
      });
    });

    context("when process.cwd() is missing package.json", function() {
      before(function() {
        this.cwd = process.cwd();

        process.chdir(__dirname);
      });

      after(function() {
        process.chdir(this.cwd);
      });

      it("should throw", function() {
        expect(function() {
          installer.check("anything");
        }).toThrow(/Cannot find module/);
      });
    });

    context("given a dependency in package.json", function() {
      it("should return undefined", function() {
        expect(installer.check("cross-spawn")).toBe(undefined);
      });
    });

    context("given a devDependency in package.json", function() {
      it("should return undefined", function() {
        expect(installer.check("expect")).toBe(undefined);
      });
    });

    context("given a linked dependency", function() {
      beforeEach(function() {
        this.lstatSync = expect.spyOn(fs, "lstatSync").andReturn({
          isSymbolicLink: function() {
            return true;
          },
        });
      });

      afterEach(function() {
        this.lstatSync.restore();
      });

      it("should return undefined", function() {
        expect(installer.check("something-linked")).toBe(undefined);
        expect(this.lstatSync.calls.length).toBe(1);
        expect(this.lstatSync.calls[0].arguments).toEqual([
          [process.cwd(), "node_modules", "something-linked"].join("/"),
        ]);
      });
    });

    context("given a global module", function() {
      it("should return undefined", function () {
        expect(installer.check("path")).toBe(undefined);
      });
    });

    context("given a module", function() {
      it("should return module", function() {
        expect(installer.check("react")).toBe("react");
      });
    });

    context("given a module/and/path", function() {
      it("should return module", function() {
        expect(installer.check("react/proptypes")).toBe("react");
      });
    });

    context("given a @namespaced/module", function() {
      it("should return @namespaced/module", function() {
        expect(installer.check("@namespaced/module")).toBe("@namespaced/module");
      });
    });

    context("given a module already installed, but not saved", function() {
      it("should return module", function() {
        expect(installer.check("yargs")).toBe("yargs");
      });
    });

    context("given a webpack !!loader/module", function() {
      it("should return undefined", function() {
        expect(installer.check("!!./css-loader/index.js',")).toBe(undefined);
      });
    })

  });

  describe(".checkBabel", function() {
    beforeEach(function() {
      this.sync = expect.spyOn(spawn, "sync").andReturn({ stdout: null });

      expect.spyOn(console, "info");
    });

    afterEach(function() {
      expect.restoreSpies();
    });

    context("when .babelrc doesn't exist", function() {
      beforeEach(function() {
        process.chdir(path.join(process.cwd(), "test"));
      });

      afterEach(function() {
        process.chdir(path.join(process.cwd(), ".."));
      });

      it("should return early", function() {
        var result = installer.checkBabel();

        expect(result).toBe(undefined);
        expect(this.sync).toNotHaveBeenCalled();
      });
    });

    context("when .babelrc exists", function() {
      beforeEach(function() {
        process.chdir(path.join(process.cwd(), "example/webpack1"));

        this.check = expect.spyOn(installer, "check").andCall(function(dep) {
          return dep;
        });

        this.install = expect.spyOn(installer, "install");
      });

      afterEach(function() {
        process.chdir(path.join(process.cwd(), "../../"));
      });

      it("should check plugins & presets", function() {
        installer.checkBabel();

        var checked = this.check.calls.map(function(call) {
          return call.arguments[0];
        });

        expect(this.check).toHaveBeenCalled();
        expect(this.check.calls.length).toEqual(6);
        expect(checked).toEqual([
          'babel-core',
          'babel-plugin-react-html-attrs',
          'babel-preset-react',
          'babel-preset-es2015',
          'babel-preset-stage-0',
          'babel-preset-react-hmre'
        ]);
      });

      it("should install missing plugins & presets", function() {
        installer.checkBabel();

        expect(this.install).toHaveBeenCalled();
        expect(this.install.calls.length).toEqual(1);
        expect(this.install.calls[0].arguments).toEqual([
          [
            'babel-core',
            'babel-plugin-react-html-attrs',
            'babel-preset-react',
            'babel-preset-es2015',
            'babel-preset-stage-0',
            'babel-preset-react-hmre'
          ],
        ]);
      });
    });
  });

  describe(".checkESLint", function() {
    var readFileSync = fs.readFileSync;

    var mockReadFileSync = function(expected, actual) {
      expect.spyOn(fs, "readFileSync").andCall(function(file, enc) {
        if (file.match(expected)) {
          return readFileSync(path.join(__dirname, actual), enc);
        }

        return readFileSync(file, enc);
      });
    };

    beforeEach(function() {
      this.check = expect.spyOn(installer, "check").andCall(function(dep) {
        return dep;
      });

      this.install = expect.spyOn(installer, "install");
    });

    afterEach(function() {
      expect.restoreSpies();
    });

    context("when nothing is extended", function() {
      beforeEach(function() {
        mockReadFileSync(/\.eslintrc$/, "./fixtures/eslintrc.empty.json");
      });

      it("should not install anything", function() {
        installer.checkESLint();

        expect(this.check).toNotHaveBeenCalled();
        expect(this.install).toNotHaveBeenCalled();
      });
    });

    context("when airbnb is extended", function() {
      beforeEach(function() {
        mockReadFileSync(/\.eslintrc$/, "./fixtures/eslintrc.airbnb.json");
      });

      afterEach(function() {
        expect.restoreSpies();
      });

      it("should .check eslint-config-airbnb", function() {
        installer.checkESLint();

        expect(this.check).toHaveBeenCalled();
        expect(this.check.calls.length).toEqual(1);
        expect(this.check.calls[0].arguments).toEqual(["eslint-config-airbnb"]);
      });

      it("should .install eslint-config-airbnb + peers as dev", function() {
        installer.checkESLint();

        expect(this.install).toHaveBeenCalled();
        expect(this.install.calls.length).toEqual(1);
        expect(this.install.calls[0].arguments).toEqual([
          ["eslint-config-airbnb"],
          {
            dev: true,
            peerDependencies: true,
          }
        ]);
      });
    });

    context("when eslint:recommended is extended", function() {
      beforeEach(function() {
        mockReadFileSync(/\.eslintrc$/, "./fixtures/eslintrc.recommended.json");
      });

      afterEach(function() {
        expect.restoreSpies();
      });

      it("should .check eslint-config-eslint", function() {
        installer.checkESLint();

        expect(this.check).toHaveBeenCalled();
        expect(this.check.calls.length).toEqual(1);
        expect(this.check.calls[0].arguments).toEqual(["eslint-config-eslint"]);
      });
    });

    context("when multiple configs are extended", function() {
      beforeEach(function() {
        mockReadFileSync(/\.eslintrc$/, "./fixtures/eslintrc.array.json");
      });

      afterEach(function() {
        expect.restoreSpies();
      });

      it("should .install eslint-config-airbnb", function() {
        installer.checkESLint();

        expect(this.install).toHaveBeenCalled();
        expect(this.install.calls.length).toEqual(1);
        expect(this.install.calls[0].arguments[0]).toEqual([
          "eslint-config-eslint",
          "eslint-config-airbnb",
        ]);
      });
    });

    context("when @highereducation/eslint-config is extended", function() {
      beforeEach(function() {
        mockReadFileSync(/\.eslintrc$/, "./fixtures/eslintrc.namespace.json");
      });

      afterEach(function() {
        expect.restoreSpies();
      });

      it("should .check @highereducation/eslint-config", function() {
        installer.checkESLint();

        expect(this.check).toHaveBeenCalled();
        expect(this.check.calls.length).toEqual(1);
        expect(this.check.calls[0].arguments).toEqual(["@highereducation/eslint-config"])
      });
    });
  });

  describe(".checkPackage", function() {
    beforeEach(function() {
      this.sync = expect.spyOn(spawn, "sync").andReturn({ stdout: null });

      expect.spyOn(console, "info");
    });

    afterEach(function() {
      expect.restoreSpies();
    });

    context("when package.json exists", function() {
      it("should return early", function() {
        var result = installer.checkPackage();

        expect(result).toBe(undefined);
        expect(this.sync).toNotHaveBeenCalled();
      });
    });

    context("when package.json does not exist", function() {
      beforeEach(function() {
        process.chdir(path.join(process.cwd(), "test"));
      });

      afterEach(function() {
        process.chdir(path.join(process.cwd(), ".."));
      });

      it("should initialize NPM", function() {
        installer.checkPackage();

        expect(this.sync).toHaveBeenCalled();
        expect(this.sync.calls.length).toEqual(1);
        expect(this.sync.calls[0].arguments).toEqual([
          "npm",
          ["init -y"],
          { stdio: "inherit" },
        ]);
      });
    });
  });

  describe(".install", function() {
    beforeEach(function() {
      this.sync = expect.spyOn(spawn, "sync").andReturn({ stdout: null });

      expect.spyOn(console, "info");
      expect.spyOn(console, "warn");
    });

    afterEach(function() {
      expect.restoreSpies();
    });

    context("given a falsey value", function() {
      it("should return undefined", function() {
        expect(installer.install()).toEqual(undefined);
        expect(installer.install(0)).toEqual(undefined);
        expect(installer.install(false)).toEqual(undefined);
        expect(installer.install(null)).toEqual(undefined);
        expect(installer.install("")).toEqual(undefined);
      });
    });

    context("given an empty array", function() {
      it("should return undefined", function () {
        expect(installer.install([])).toEqual(undefined);
      });
    });

    context("given a non-existant module", function() {
      beforeEach(function() {
        this.sync.andReturn({ status: 1 });
      });

      it("should attempt to install once", function() {
        installer.install("does.not.exist.jsx")

        expect(this.sync).toHaveBeenCalled();
      });

      it("should not attempt to install it again", function() {
        installer.install("does.not.exist.jsx")

        expect(this.sync).toNotHaveBeenCalled();
      });
    });

    context("given a dependency", function() {
      context("with no options", function() {
        it("should install it with --save", function() {
          var result = installer.install("foo");

          expect(this.sync).toHaveBeenCalled();
          expect(this.sync.calls.length).toEqual(1);
          expect(this.sync.calls[0].arguments[0]).toEqual("npm");
          expect(this.sync.calls[0].arguments[1]).toEqual(["install", "foo", "--save"]);
        });
      });

      context("with dev set to true", function() {
        it("should install it with --save-dev", function() {
          var result = installer.install("foo", {
            dev: true,
          });

          expect(this.sync).toHaveBeenCalled();
          expect(this.sync.calls.length).toEqual(1);
          expect(this.sync.calls[0].arguments[0]).toEqual("npm");
          expect(this.sync.calls[0].arguments[1]).toEqual(["install", "foo", "--save-dev"]);
        });
      });

      context("with missing peerDependencies", function() {
        beforeEach(function() {
          this.sync.andCall(function(bin, args) {
            var dep = args[1];

            if (dep === "redbox-react") {
              return {
                stdout: new Buffer([
                  "/test",
                  "├── redbox-react@1.2.3",
                  "└── UNMET PEER DEPENDENCY react@>=0.13.2 || ^0.14.0-rc1 || ^15.0.0-rc",
                ].join("\n")),
              };
            }

            return { stdout: null };
          });
        });

        context("given no options", function() {
          it("should install peerDependencies", function() {
            var result = installer.install("redbox-react");

            expect(this.sync.calls.length).toEqual(2);
            expect(this.sync.calls[0].arguments[1]).toEqual(["install", "redbox-react", "--save"]);
            expect(this.sync.calls[1].arguments[1]).toEqual(["install", "react@\">=0.13.2 || ^0.14.0-rc1 || ^15.0.0-rc\"", "--save"]);
          });
        });

        context("given peerDependencies set to false", function() {
          it("should not install peerDependencies", function() {
            var result = installer.install("redbox-react", {
              peerDependencies: false,
            });

            expect(this.sync.calls.length).toEqual(1);
            expect(this.sync.calls[0].arguments[1]).toEqual(["install", "redbox-react", "--save"]);
          });
        });
      });
    });
  });
});
