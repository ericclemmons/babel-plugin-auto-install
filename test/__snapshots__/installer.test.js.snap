// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`installer .install when using npm given a dependency with missing peerDependencies given dependencies.peer set to false should not install peerDependencies 1`] = `
Array [
  Array [
    "npm",
    Array [
      "install",
      "redbox-react",
      "--save",
    ],
    Object {
      "stdio": Array [
        "ignore",
        "pipe",
        "inherit",
      ],
    },
  ],
]
`;

exports[`installer .install when using pnpm given a dependency with missing peerDependencies given peerDependencies set to false should not install peerDependencies 1`] = `
Array [
  Array [
    "pnpm",
    Array [
      "add",
      "redbox-react",
    ],
    Object {
      "stdio": Array [
        "ignore",
        "pipe",
        "inherit",
      ],
    },
  ],
]
`;

exports[`installer .install when using yarn given a dependency with missing peerDependencies given peerDependencies set to false should not install peerDependencies 1`] = `
Array [
  Array [
    "yarn",
    Array [
      "add",
      "redbox-react",
    ],
    Object {
      "stdio": Array [
        "ignore",
        "pipe",
        "inherit",
      ],
    },
  ],
]
`;
